'Introducing Bloom Filters for Valkey' Blog (#229)
valkey-io/valkey-io.github.io: content/blog/2025-04-09-introducing-bloom-filters.md

```
127.0.0.1:6379> BF.INFO filter1 MAXSCALEDCAPACITY
(integer) 26214300
```

To get an idea of what the memory usage looks like for the max capacity of an individual non scaling filter, we have a table below.
With a 128MB limit and default false positive rate, we can create a bloom filter with 112M as the capacity. With a 512MB limit, a bloom filter can hold 448M items.

| Non Scaling Filter - Capacity | FP Rate | Memory Usage (MB) | Notes                                    |
|------------------------------|---------|-------------------|------------------------------------------|
| 112M                         | 0.01    | ~128              | Default FP Rate and Default Memory Limit |
| 74M                          | 0.001   | ~128              | Custom FP Rate and Default Memory Limit |
| 448M                         | 0.01    | ~512              | Default FP Rate and Custom Memory Limit |
| 298M                         | 0.001   | ~512              | Custom FP Rate and Custom Memory Limit  |

## Performance 

> The bloom commands which involve adding items or checking the existence of items have a time complexity of O(N * K) where N is the number of hash functions used by the bloom filter and K is the number of elements being inserted.
> This means that both `BF.ADD` and `BF.EXISTS` are both O(N) as they only operate on one item.

Author: ```suggestion
The bloom commands which involve adding items or checking the existence of items have a time complexity of O(N * K) where N is the number of elements being inserted and K is the number of hash functions used by the bloom filter.
This means that both `BF.ADD` and `BF.EXISTS` are both O(K) as they only operate on one item.
```

+++
title= "Introducing Bloom Filters for Valkey"
description = "Learn how to use bloom filters to perform large-scale membership testing with significant memory savings."
date= 2025-04-09 01:01:01
authors= [ "karthiksubbarao"]
+++

The Valkey project is introducing Bloom Filters as a new data type via [valkey-bloom](https://github.com/valkey-io/valkey-bloom/), an official Valkey Module which is compatible with Valkey versions >= 8.0. Bloom filters provide efficient, large-scale membership testing, improving performance and offering significant memory savings for high-volume applications.

> As an example, to handle advertisement deduplication workloads and answer the question, "Has this customer seen this ad before?", previously, Valkey developers could use the SET data type.
> This is done by adding the customer IDs (of those who viewed an ad) into a `SET` object representing a particular advertisement. However, since every item in the set is allocated, the problem with this approach is high memory usage.
> This article demonstrates how using the bloom filter data type from valkey-bloom achieves >93% memory savings for the same workload, while exploring its implementation, technical details, and practical recommendations.

Reviewer: ```suggestion
As an example, to handle advertisement deduplication workloads and answer the question, "Has this customer seen this ad before?", Valkey developers could use the SET data type.
This is done by adding the customer IDs (of those who viewed an ad) into a `SET` object representing a particular advertisement. However, the problem with this approach is high memory usage since every item in the set is allocated.
This article demonstrates how using the bloom filter data type from valkey-bloom can achieve significant memory savings, more than 93% in our example workload, while exploring its implementation, technical details, and practical recommendations.
```
Tried to clean up some of the wording and move to more active voice for some earlier paragraphs. Didn't think it was as important for later in the blog.

## Introduction

> Bloom filters are a space efficient probabilistic data structure that allows adding elements and checking whether elements exist. False positives are possible where a filter incorrectly indicates that an element exists, even though it was not added.

Reviewer: ```suggestion
Bloom filters are a space efficient probabilistic data structure that supports adding elements and checking whether elements were previously added. False positives are possible, where a filter incorrectly indicates that an element exists even though it was not added.
```

> However, Bloom Filters guarantee that false negatives do not occur, meaning it can never be the case that an element was added successfully and a filter reports it as not existing.

Reviewer: ```suggestion
However, Bloom Filters guarantee that false negatives do not occur, meaning that an element that was added successfully will never be reported as not existing.
```

> Bloom Filters were first introduced in a paper from 1970 by Burton H. Bloom.

Reviewer: ```suggestion
```
Do we need this, feels like a tangent?

![Bloom Filter Bit Vector](/assets/media/pictures/bloomfilter_bitvector.png)
*Image taken from [source](https://en.wikipedia.org/wiki/Bloom_filter#/media/File:Bloom_filter.svg)*

> When adding an item to a bloom filter, K hash functions compute K corresponding bits from the bit vector which are set to 1.

Reviewer: ```suggestion
When adding an item to a bloom filter, some number, K, hash functions compute K corresponding bits from the bit vector which are set to 1.
```

Author: How about this?
> When adding an item to a bloom filter, K different hash functions compute K corresponding bits from the bit vector which are set to 1.

Checking existence involves the same hash functions - if any bit is 0, the item is definitely absent; if all bits are 1, the item likely exists (with a defined false positive probability).
This bit-based approach, rather than full item allocation, makes bloom filters very space efficient with the trade off being potential false positives.

> Valkey-Bloom (BSD-3 licensed) is an official Valkey module that introduces bloom filters as a new data type to Valkey, providing both scalable and non-scalable variants.

Reviewer: ```suggestion
Valkey-Bloom introduces bloom filters as a new data type to Valkey, providing both scalable and non-scalable variants.
```
We already introduce it as an official module earlier. I don't think we need to repeat it.

Author: Yes, I can remove that part of the sentence here. However, we did not mention the license earlier. Should I move that to the introduction paragraph or is the BSD3 license not important to mention?

It is API compatible with the bloom filter command syntax of the official Valkey client libraries including valkey-py, valkey-java, valkey-go (as well as the equivalent Redis libraries).

> The module was developed by the following authors: Karthik Subbarao ([KarthikSubbarao](https://github.com/KarthikSubbarao)), Cameron Zack ([zackcam](https://github.com/zackcam)), Vanessa Tang ([YueTang-Vanessa](https://github.com/YueTang-Vanessa)), Nihal Mehta ([nnmehta](https://github.com/nnmehta)), and wuranxx ([wuranxx](https://github.com/wuranxx)).

Reviewer: I would move the callout to the community to the bottom. I think it's important to include, but it breaks the flow for readers.

![bloom filter data type](/assets/media/pictures/bloomfilter_datatype.png)

**Non Scaling**

When non-scaling filters reach their capacity, if a user tries to add a new/unique item, an error is returned.
You can create a non scaling bloom filter using `BF.RESERVE` or `BF.INSERT` commands.
Example:
```
BF.RESERVE <filter-name> <error-rate> <capacity> NONSCALING
``` 

**Scaling**

When scaling filters reach their capacity, if a user adds an item to the bloom filter, a new sub filter is created and added to the vector of sub filters.
This new bloom sub filter will have a larger capacity (previous_bloomfilter_capacity * expansion_rate of the bloom filter).
When checking whether an item exists on a scaled out bloom filter (`BF.EXISTS`/`BF.MEXISTS`), we look through each filter (from oldest to newest) in the sub filter vector and perform a check operation on each one.
Similarly, to add a new item to the bloom filter, we check through all the filters to see if the item already exists and the item is added to the current filter if it does not exist.
Any default creation as a result of `BF.ADD`, `BF.MADD`, `BF.INSERT` will be a scalable bloom filter.

**Common Bloom filter properties**

> Capacity - The number of unique items that can be added before a bloom filter scales out occurs (in case of scalable bloom filters) or before any command which inserts a new item is rejected (in case of non scalable bloom filters).

Reviewer: ```suggestion
Capacity - The number of unique items that can be added before a bloom filter scales out occurs (in case of scalable bloom filters) or before any command which inserts a new item will return an error (in case of non scalable bloom filters).
```

| Number of Bloom Filters | Capacity | FP Rate | FP Rate Description  | Total Used Memory (GB) | Memory Saved % compared to SETS |
|-------------------------|----------|---------|----------------------|------------------------|----------------------------------|
| 500                     | 5000000  | 0.01    | One in every 100      | 2.9                    | **98.08%** |
| 500                     | 5000000  | 0.001   | One in every 1K       | 4.9                    | **96.80%** |
| 500                     | 5000000  | 0.00001 | One in every 100K     | 7.8                    | **94.88%** |
| 500                     | 5000000  | 0.0000002| One in every 5M      | 9.8                    | **93.60%** |

In this example, we are able to benefit from 93% - 98% savings in memory usage when using Bloom Filters compared to the SET data type. Depending on your workload, you can expect similar results.

![SET vs Bloom Filter Memory Usage Comparison](/assets/media/pictures/bloomfilter_memusage.png)

## Large Bloom Filters and Recommendations

> To ensure optimal server performance during serialization and deserialization of bloom filters, we have added validation on the memory usage per object.

Reviewer: ```suggestion
To improve server performance during serialization and deserialization of bloom filters, we have added validation on the memory usage per object.
```
We can't ensure anything.

> The default memory usage limit of a bloom filter is defined by the `BF.BLOOM-MEMORY-USAGE-LIMIT` module configuration which has a default value of 128 MB.

Reviewer: ```suggestion
The default memory usage limit of a bloom filter is defined by the `BF.BLOOM-MEMORY-USAGE-LIMIT` configuration which has a default value of 128 MB.
```

However, the value can be tuned using the configuration above.

> The implication of the memory limit is that operations involving bloom filter creations or scaling out that result in a bloom filter with overall memory usage over the limit will be rejected. Example:

Reviewer: ```suggestion
The implication of the memory limit is that operations involving bloom filter creations or scaling out that result in a bloom filter with overall memory usage over the limit will return an error. Example:
```

```
127.0.0.1:6379> BF.ADD ad1_filter user1
(error) ERR operation exceeds bloom object memory limit
```
This poses an issue to users where their scalable bloom filters can reach the memory limit after some number of days of data population and it starts failing scale outs during the insertion of unique items.

As a solution, to help users understand at what capacity their bloom filter will hit the memory limit, valkey-bloom has two options.
These are useful to check beforehand to ensure that your bloom filter will not fail scale outs or creations later on as part of your workload.

1. Perform a memory check prior to bloom filter creation

We can use the `VALIDATESCALETO` option of the `BF.INSERT` command to perform a validation whether the filter is within the memory limit.

> If it is not within the limits, the command is rejected. In the example below, we see that filter1 cannot scale out and reach the capacity of 26214301 due to the memory limit. However, it can scale out and reach a capacity of 26214300.

Reviewer: ```suggestion
If it is not within the limits, the command will return an error. In the example below, we see that filter1 cannot scale out and reach the capacity of 26214301 due to the memory limit. However, it can scale out and reach a capacity of 26214300.
```

```
127.0.0.1:6379> BF.INFO filter1 MAXSCALEDCAPACITY
(integer) 26214300
```

To get an idea of what the memory usage looks like for the max capacity of an individual non scaling filter, we have a table below.
With a 128MB limit and default false positive rate, we can create a bloom filter with 112M as the capacity. With a 512MB limit, a bloom filter can hold 448M items.

| Non Scaling Filter - Capacity | FP Rate | Memory Usage (MB) | Notes                                    |
|------------------------------|---------|-------------------|------------------------------------------|
| 112M                         | 0.01    | ~128              | Default FP Rate and Default Memory Limit |
| 74M                          | 0.001   | ~128              | Custom FP Rate and Default Memory Limit |
| 448M                         | 0.01    | ~512              | Default FP Rate and Custom Memory Limit |
| 298M                         | 0.001   | ~512              | Custom FP Rate and Custom Memory Limit  |

## Performance 

> The bloom commands which involve adding items or checking the existence of items have a time complexity of O(N * K) where N is the number of hash functions used by the bloom filter and K is the number of elements being inserted.

Reviewer: ```suggestion
The bloom commands which involve adding items or checking the existence of items have a time complexity of O(N * K) where N is the number of hash functions used by the bloom filter and K is the number of elements being inserted.
```
Earlier in the document you said K is number of hash functions, might be better to be consistent with that. You also say O(N) in the next sentence, which confused me at first since I assumed N was number of elements in the system.

Author: Thanks - yes, I noticed too earlier and forgot about changing this. I will update the sentence

This means that both `BF.ADD` and `BF.EXISTS` are both O(N) as they only operate on one item.

> In scalable bloom filters, with every scale out, we increase the number of hash function based checks during add/exists operations; Each sub filter requires at least one hash function and this number increases as the false positive rate becomes stricter with scale outs due to the [tightening ratio](https://valkey.io/topics/bloomfilters/#advanced-properties).

Reviewer: ```suggestion
In scalable bloom filters, we increase the number of hash function based checks during add/exists operations with each scale out; Each sub filter requires at least one hash function and this number increases as the false positive rate becomes stricter with scale outs due to the [tightening ratio](https://valkey.io/topics/bloomfilters/#advanced-properties).
```
The two clauses just reads weird to me. Minor suggestion.
